firmware-support/bittide-sys/src/sample_store.rs

```diff
@@ -0,0 +1,81 @@
+// SPDX-FileCopyrightText: 2025 Google LLC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use bittide_hal::freeze::SampleMemory;
+
+const WORDS_PER_SAMPLE: usize = 11;
+
+/// State machinery for storing clock control samples in memory.
+pub struct SampleStore {
+    memory: SampleMemory,
+    store_samples_every: usize,
+    counter: usize,
+}
+
+impl SampleStore {
+    pub fn new(memory: SampleMemory, store_samples_every: usize) -> Self {
+        // First memory location is reserved for the number of samples stored.
+        memory.set_data(0, 0);
+
+        Self {
+            memory,
+            store_samples_every,
+            counter: 0,
+        }
+    }
+
+    /// *Actually* store the contents of 'Freeze' to memory. Note that the public
+    /// function 'store' does 'store_samples_every' boundary checking.
+    fn do_store(&mut self, freeze: &bittide_hal::freeze::Freeze, bump_counter: bool) {
+        let n_samples_stored = self.memory.data(0).unwrap() as usize;
+        let start_index = n_samples_stored * WORDS_PER_SAMPLE + 1;
```

Comment:
In retrospect: could we have had a block ram that has a fancy record type instead?

Reply:
Not really.. I'll write up an issue.